### ISSUE TITLE
 Ensuring gradient flow from `flax` neural network output to `synthax` synth parameters

### ISSUE THREAD

**Issue by jofrevalles** [PapayaResearch/synthax#1]
Hello, `synthax` team!

I'm working on a project where I use `synthax` to generate audio based on parameters predicted by a neural network (NN) using `flax`. The NN predicts a set of parameters that should be passed into the synth using `jax.jit(synth.apply)(params)` to create audio. 

However, I've encountered a challenge regarding the gradient flow from the network predictions to the `synthax` synthesizer parameters. The goal is to backpropagate through the synthesizer's parameter updates to optimize the NN's predictions. To achieve this, I'm attempting to maintain the gradient information from the NN's output as it's transformed and applied to the synthesizer.

My primary concern is ensuring that the transformation process from the output of the NN to the `FrozenDict` maintains the gradient information so that the NN can be effectively trained based on the audio output.

Could you provide guidance or best practices on how to ensure that the gradient flow is preserved during this transformation process? Specifically, I'm looking for advice on how to structure this transformation in a way that is compatible with `JAX`'s automatic differentiation, ensuring that gradients can propagate back through the parameter transformation step to the NN. I see that the structure of `flax` networks and the synths in `synthax` are similar so I think I am missing something.

Any insights or recommendations would be greatly appreciated. 

Thank you!
Jofre

**Comment by nikhilsinghmus** [comment]
Hi @jofrevalles,

Thanks for your comment! If I understand correctly, you're working on transforming the output from a model (an array of shape `(batch_size, params)`) into the parameter dict, to drive a Synthax synthesizer? If so, here's a very quick toy example. Let us know if you already considered something like this:

```python
import jax
import jax.flatten_util
import jax.numpy as jnp
import flax.linen as nn
from synthax.config import SynthConfig
from synthax.synth import Voice


batch_size = 16
n_features = 300

class SimpleMLP(nn.Module):
    features: int

    @nn.compact
    def __call__(self, x):
        x = nn.Dense(features=self.features)(x)
        x = nn.relu(x)
        x = nn.Dense(features=78)(x)
        return x

config = SynthConfig(
    batch_size=batch_size,
    sample_rate=44100,
    buffer_size_seconds=4.0
)

# Instantiate and initialize synthesizer
synth = Voice(config=config)
synth_apply = jax.jit(synth.apply)
key = jax.random.PRNGKey(42)
init_params = synth.init(key)

# Get n_params for ravel/unravel
init_params = jax.tree_util.tree_map(lambda x : x[0], init_params)
_, unravel_pytree = jax.flatten_util.ravel_pytree(init_params)

# vmap for batch dimension
unravel_pytree = jax.vmap(unravel_pytree)

# Instantiate MLP
mlp = SimpleMLP(features=n_features)
mlp_params = mlp.init(key, jnp.ones((batch_size, n_features)))
mlp_apply = jax.jit(mlp.apply)

# Predict params and synthesize
def predict_params_and_synthesize(x):
    params = mlp_apply(mlp_params, x)
    params = nn.sigmoid(params) # Sigmoid to ensure params are in [0, 1]
    params = unravel_pytree(params)
    return synth_apply(params)

control_input = jax.random.normal(
    jax.random.split(key)[0],
    (batch_size, n_features)
)

audio = predict_params_and_synthesize(control_input)
```

It's worth noting that the gradients from an architecture like `Voice` with respect to the parameters can be very unstable and difficult to optimize. For example, in our [`ctag`](https://github.com/PapayaResearch/ctag) project, we used [LES](https://github.com/RobertTLange/evosax/blob/main/evosax/strategies/les.py) as an optimizer with this architecture.

**Comment by jofrevalles** [comment]
Thank you so much for the quick answer, @nikhilsinghmus!

I did not know about this `jax.flatten_util.ravel_pytree` notation, this makes a lot of sense! Sorry, I am still not completely familiar with the `jax` notation. This example is perfect for what I am looking for.

On a side note, I am trying this inspired from your `ctag` project, which I found very interesting. 
Thank you for the example and the tips! I am closing the issue.
